% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/make_mut_matrix.R
\name{make_mut_matrix}
\alias{make_mut_matrix}
\title{Make mutation matrix using maf data}
\usage{
make_mut_matrix(
  maf = NULL,
  mut_data = NULL,
  isTCGA = TRUE,
  category = "multi",
  Tumor_Sample_Barcode = "Tumor_Sample_Barcode",
  Hugo_Symbol = "Hugo_Symbol",
  Variant_Classification = "Variant_Classification",
  Variant_Type = "Variant_Type"
)
}
\arguments{
\item{maf}{file name or a maf object}

\item{mut_data}{if maf data had beed loaded, this argument will be applied}

\item{isTCGA}{logical variable,}

\item{category}{if category is `multi`, four mutation matrix will be generated: all, snp, indel and frameshift}

\item{Tumor_Sample_Barcode}{column name of tumor sample}

\item{Hugo_Symbol}{column name of gene symbol}

\item{Variant_Classification}{column name of variant classification: Frame_Shift_Del}

\item{Variant_Type}{column name of variant type: SNP, INS, DEL}
}
\value{
A list of matrices, each representing a different category of mutations as specified by `category`.
}
\description{
This function constructs mutation matrices from MAF data, supporting multiple categories including all mutations,
SNPs, indels, and frameshift mutations. It can handle both MAF file paths or MAF objects directly and is capable
of processing data formatted according to TCGA standards.
}
\examples{

library(TCGAbiolinks)
query <- GDCquery(project = "TCGA-STAD", data.category = "Simple Nucleotide Variation", access = "open", data.type = "Masked Somatic Mutation", workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking" )
GDCdownload(query)
maf <- GDCprepare(query)
mut_list <- make_mut_matrix(maf = maf, isTCGA   = T, category = "multi")
}
\author{
Dongqian Zeng

Shixiang Huang
}
